### PR TITLE
Added support for transition from resourcepool to subnamespace

### LINKED ISSUE #23: add support for going back from resourcepool to subnamespace
At the moment it's only possible to switch from "regular" subnamespace to ResourcePool, but going the other way does not work properly. Changing the top-level ResourcePool works fine but its children are updated without a limit (subnamespace spec is empty). 

We should discuss how we want to handle it and fix it.

/kind feature

### PR DESCRIPTION
**Pull request by omerbd21**
Fixes #23. With this PR I added a trigger to child subnamespace if it is a resourcepool who has is-upper-rp=False and its father is a subnamespace. In addition, I added a check before the updateObject in the sync function of the subnamespace controller, so that the update will occur only if needed.

### CODE REVIEW

File: internals/controllers/namespace_controller.go [hunk 1]
```diff
@@ -293,9 +294,9 @@ func ensureSnsResourcePool(namespace *utils.ObjectContext) error {
 			return err
 		}
 
-		if utils.GetSnsResourcePooled(snsToUpdate.Object) != namespaceResourcePooled && !utils.IsRootResourcePool(snsToUpdate) {
+		if utils.GetNamespaceResourcePooled(namespace) == "true" && !utils.IsRootResourcePool(snsToUpdate) {
```
Comment (mzeevi) [inline internals/controllers/namespace_controller.go:297]: I'm unsure we even need the resourcepool label on the namespace, do we even do anything with it?

File: internals/controllers/subnamespace_controller.go [hunk 2]
```diff
@@ -125,6 +125,34 @@ func (r *SubnamespaceReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
+	return true
+}
+func resourceListEqual(a, b corev1.ResourceList) bool {
+	if b.Cpu().AsApproximateFloat64() != a.Cpu().AsApproximateFloat64() ||
+		b.Memory().AsApproximateFloat64() != a.Memory().AsApproximateFloat64() ||
+		b.Pods().AsApproximateFloat64() != a.Pods().AsApproximateFloat64() {
+		return false
+	}
+	return true
+}
```
Comment (mzeevi) [inline internals/controllers/subnamespace_controller.go:155]: Add a // comment for every function explaining what it does

File: internals/controllers/subnamespace_controller.go [hunk 3]
```diff
@@ -177,15 +205,22 @@ func (r *SubnamespaceReconciler) Sync(ownerNamespace *utils.ObjectContext, subsp
...
-			},
-		}}
+	for _, sns := range subspaceChilds.Objects.(*danav1.SubnamespaceList).Items {
+		snsChild, err := utils.NewObjectContext(subspace.Ctx, subspace.Log, subspace.Client, types.NamespacedName{Name: sns.GetName(), Namespace: subspace.GetName()}, &danav1.Subnamespace{})
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if utils.GetSnsResourcePooled(subspace.Object) == "false" &&
+			utils.GetSnsResourcePooled(snsChild.Object) == "true" &&
+			snsChild.Object.GetAnnotations()[danav1.IsUpperRp] == danav1.False {
```
Comment (mzeevi) [inline internals/controllers/subnamespace_controller.go:215]: This `if` statement is way to convoluted, move it to a function with a relevant name that returns a boolean; will make this more readable as well

File: internals/controllers/subnamespace_controller.go [hunk 4]
```diff
@@ -177,15 +205,22 @@ func (r *SubnamespaceReconciler) Sync(ownerNamespace *utils.ObjectContext, subsp
...
+			r.SnsEvents <- event.GenericEvent{Object: &danav1.Subnamespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      sns.GetName(),
+					Namespace: subspace.Object.GetName(),
+				},
```
Comment (mzeevi) [inline internals/controllers/subnamespace_controller.go:220]: Add a comment explaining why we need to trigger reconciliation for the children of the ResourcePool that got changed

File: internals/controllers/subnamespace_controller.go [hunk 5]
```diff
@@ -225,20 +259,31 @@ func (r *SubnamespaceReconciler) Sync(ownerNamespace *utils.ObjectContext, subsp
 			return ctrl.Result{}, err
 		}
 	}
-
+	if _, ok := subspace.Object.GetAnnotations()[danav1.IsRq]; !ok {
+		if err = subspace.AppendAnnotations(map[string]string{danav1.IsRq: danav1.False}); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
```
Comment (mzeevi) [inline internals/controllers/subnamespace_controller.go:266]: This adds `isR` to ResourcePools as well?

File: internals/controllers/subnamespace_controller.go [hunk 6]
```diff
@@ -258,13 +303,17 @@ func (r *SubnamespaceReconciler) Sync(ownerNamespace *utils.ObjectContext, subsp
...
-	if err := subspace.UpdateObject(func(object client.Object, log logr.Logger) (client.Object, logr.Logger) {
-		object.(*danav1.Subnamespace).Status.Namespaces = childrenRequests
-		object.(*danav1.Subnamespace).Status.Total.Allocated = allocated
-		object.(*danav1.Subnamespace).Status.Total.Free = free
-		return object, log
-	}); err != nil {
-		return ctrl.Result{}, err
+	if !namespacesEqual(subspace.Object.(*danav1.Subnamespace).Status.Namespaces, childrenRequests) ||
+		!resourceListEqual(subspace.Object.(*danav1.Subnamespace).Status.Total.Allocated, allocated) ||
+		!resourceListEqual(subspace.Object.(*danav1.Subnamespace).Status.Total.Free, free) {
```
Comment (mzeevi) [inline internals/controllers/subnamespace_controller.go:308]: Move this entire if statement to a boolean function with a relevant name

File: internals/controllers/subnamespace_controller.go [hunk 7]
```diff
@@ -663,6 +711,7 @@ func composeChildNamespace(ownerNamespace *utils.ObjectContext, subspace *utils.
 
 	ann := getParentAnnotations(ownerNamespace)
 	ann[danav1.Depth] = childNamespaceDepth
+	ann[danav1.IsRq] = danav1.False
```
Comment (mzeevi) [inline internals/controllers/subnamespace_controller.go:714]: Is this on purpose?

File: internals/utils/helper.go [hunk 8]
```diff
@@ -62,6 +62,9 @@ func AppendUpperResourcePoolAnnotation(sns *ObjectContext, parentSns *ObjectCont
 		if err = sns.AppendAnnotations(map[string]string{danav1.IsUpperRp: danav1.True}); err != nil {
 			return err
 		}
+		if err = sns.DeleteAnnotations([]string{danav1.UpperRp}); err != nil {
+			return err
+		}
```
Comment (mzeevi) [inline internals/utils/helper.go:67]: I'm not sure we need to delete it? I think it would be fine a `upperRp` label to have this annotation, no?

File: internals/utils/helper.go [hunk 9]
```diff
@@ -78,7 +81,7 @@ func AppendUpperResourcePoolAnnotation(sns *ObjectContext, parentSns *ObjectCont
 		}
 	}
 	upperRp, exists := parentSns.Object.GetAnnotations()[danav1.UpperRp]
-	if exists {
+	if exists && !isUpperResourcePool {
```
Comment (mzeevi) [inline internals/utils/helper.go:84]: Why is this needed?

File: internals/utils/objectContext.go [hunk 10]
```diff
@@ -148,6 +149,22 @@ func (r *ObjectContext) AppendAnnotations(annotationsToAppend map[string]string)
 	return nil
 }
 
+func (r *ObjectContext) DeleteAnnotations(annotationsToDelete []string) error {
```
Comment (mzeevi) [inline internals/utils/objectContext.go:152]: Add a // comment for the function

File: internals/utils/objectContext.go [hunk 11]
```diff
@@ -210,7 +227,7 @@ func (r *ObjectContext) UpdateNsByparent(nsparent *ObjectContext, nsChild *Objec
 		labels[danav1.Aggragator+nsName] = "true"
 		labels[danav1.Parent] = nsparent.Object.(*corev1.Namespace).Name
 		labels[danav1.Hns] = "true"
-		labels[danav1.ResourcePool] = GetNamespaceResourcePooled(nsChild)
+		//labels[danav1.ResourcePool] = GetNamespaceResourcePooled(nsChild)
```
Comment (mzeevi) [inline internals/utils/objectContext.go:230]: Maybe we can delete this all together? Don't think we use the namespace ResourcePool label anywhere else

File: internals/webhooks/subnamespace_webhook.go [hunk 12]
```diff
@@ -162,11 +162,11 @@ func (a *SubNamespaceAnnotator) Handle(ctx context.Context, req admission.Reques
...
-		if utils.GetSnsResourcePooled(sns.Object) == "false" {
-			if isValid, err := ValidateAllResourceQuotaParamsValid(sns); !isValid {
-				return admission.Denied(err.Error())
-			}
-		}
+		// if utils.GetSnsResourcePooled(sns.Object) == "false" && utils.GetSnsResourcePooled(oldSns.Object) == "false" {
+		// 	if isValid, err := ValidateAllResourceQuotaParamsValid(sns); !isValid {
+		// 		return admission.Denied(err.Error())
+		// 	}
+		// }
```
Comment (mzeevi) [inline internals/webhooks/subnamespace_webhook.go:169]: Need to check if this is still a problem since this was put there by design

File: test/e2e/resourcepool_test.go [hunk 13]
```diff
@@ -94,7 +96,7 @@ var _ = Describe("ResourcePool", func() {
 		MustRun("kubectl delete subnamespace -n", nsC, nsD)
 	})
 
-	It("should create a clusterresourcequota for a resourcepool regardless of its depth only if it's upper", func() {
+	It("should create a clusterresourcequota for a resourcepool regardless ofP its depth only if it's upper", func() {
```
Comment (mzeevi) [inline test/e2e/resourcepool_test.go:99]: `ofP`

File: internals/controllers/namespace_controller.go (hunk 1 above)
Comment (omerbd21) [reply internals/controllers/namespace_controller.go:297]: We use it at the func ComposeChildNamespace but I also think we can delete it and not use it, it's not relevant in my opinion

File: internals/utils/helper.go (hunk 9 above)
Comment (omerbd21) [reply internals/utils/helper.go:84]: It makes sure that there won't be infinity api calls on the new upper resource pool

**Review by mzeevi — APPROVED** [review]
